src/components/EventStream.js

```diff
@@ -0,0 +1,149 @@
+
+const EventStream = ({ debugId }) => {
+  /// state ///
+  const [isFirstSnapshot, setIsFirstSnapshot] = useState(true);
+  const [pings, setPings] = useState([]);
+  const [changeQueue, setChangeQueue] = useState([]);
+
+  /// lifecycle ///
+  const onCollectionUpdate = useCallback(() => {
+    // Clear previously highlighted entries on query update.
+    let normalizedPings = pings.map((p) => {
+      p.changed = false;
+      return p;
+    });
+
+    // Iterate over all changes in change queue to add and/or remove pings.
+    changeQueue.forEach((change) => {
+      // New doc, need to append to current array.
+      if (change.type === "added") {
+        const { addedAt, payload, pingType, error, errorType, errorMessage, warning } =
+          change.doc.data();
+
+        // Add ping at the start of the array.
+        normalizedPings.unshift({
+          key: change.doc.id,
+          addedAt,
+          displayDate: formatDate(addedAt),
+          payload,
+          pingType,
+          changed: true,
+          error,
+          errorType,
+          errorMessage,
+          warning
+        });
+      }
+
+      // Doc should no longer be in our array, so we remove it.
+      if (change.type === "removed") {
+        normalizedPings = normalizedPings.filter((ping) => {
+          // Remove all pings with corresponding ID, should only ever be 1.
+          return ping.key !== change.doc.id;
+        });
+      }
+    });
+
+    // Sort the pings to show the most recent first.
+    normalizedPings.sort((a, b) => {
+      return a.addedAt > b.addedAt ? -1 : 1;
+    });
+
+    // Clear changed flag on page load to avoid whole table blinking.
+    if (isFirstSnapshot) {
+      normalizedPings.forEach((p) => {
+        p.changed = false;
+      });
+    }
+
+    setIsFirstSnapshot(false);
+    setPings(normalizedPings);
+
+    // Queued changes have been added to local pings; queue can be cleared.
+    setChangeQueue([]);
+  }, [pings, changeQueue, isFirstSnapshot]);
+
+  useEffect(() => {
+    const pingsQuery = query(
+      collection(getFirestore(), "pings"),
+      where("debugId", "==", debugId),
+      orderBy("addedAt", "desc"),
+      limit(100)
+    );
+
+    // Listener for our realtime connection to the firestore collection.
+    const unsubscribe = onSnapshot(pingsQuery, (querySnapshot) => {
+      setChangeQueue(querySnapshot.docChanges());
+    });
+
+    return () => {
+      unsubscribe();
+    };
+  }, [debugId]);
+
+  useEffect(() => {
+    if (changeQueue.length) {
+      // Process our queued changes.
+      onCollectionUpdate();
+    } else {
+      // Do nothing, we have no currently queued changes.
+    }
+  }, [changeQueue, onCollectionUpdate]);
+
+  const events = useMemo(() => {
+    let allEvents = [];
+    pings.forEach((ping) => {
+      if (ping.payload) {
+        const parsedPing = JSON.parse(ping.payload);
+        const pingEvents = parsedPing.events;
+
+        if (pingEvents) {
```

Comment:
Does having this conditional here optimize for anything when you use `forEach` on the next line?

Reply:
This conditional is just to make sure we actually have events and don't try to iterate over an undefined.

In terms of optimization, the `events` variable is wrapped in a `useMemo` so the only time `events` will ever be updated is if the contents of the `pings` variable changes, so that helps to avoid unnecessary re-renders.